### DISCUSSION TITLE
`.luarc.json` can't work with lua-language-server

### DISCUSSION THREAD

**Discussion by mrxiaozhuox** [helix-editor/helix#6651] [General]
I have installed `lua-language-server` and tried to create a Lua config file, but it's not working for the LSP server.

```
helix --healt lua

Configured language server: lua-language-server
Binary for language server: /opt/homebrew/bin/lua-language-server
Configured debug adapter: None
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```

**Comment by junelva** [comment]
In case anyone sees this unanswered question and feels discouraged, I am currently using **helix 24.7** and adding a `.luarc.json` file **works for me** here. *As an example,* here is what my file looks like. If you use it as an example, you can see the keys I used and adjust them to your liking.

```json
{
  "workspace.library": [
    "/path/to/library"
  ],
  "runtime.version": "Lua 5.2",
  "hint.enable": true
}
```

I'm using this to develop scripts that operate within the context of a lua library plugin for a game. It's cool, works fine here for me for now. Good luck!

**Reply by TurtleP** [reply]
This doesn't seem to work, and I'm also using Helix 24.7. I mainly want to use a `.luarc.json` file for code formatting, but it doesn't seem to work since it keeps giving me 2 spaces and not 4 as I have configured.

```json
{
  "$schema": "https://raw.githubusercontent.com/sumneko/vscode-lua/master/setting/schema.json",
  "runtime.version": "LuaJIT",
  "runtime.path": [
    "lua/?.lua",
    "lua/?/init.lua"
  ],
  "workspace.library": [
    "${3rd}/love2d/library"
  ],
  "hint.enable": true,
  "format.defaultConfig": {
    "indent_style": "space",
    "indent_size": "4",
    "insert_final_newline": "true",
    "space_around_table_field_list": "true",
    "align_continuous_assign_statement": "false"
  }
}
```

**Reply by mnpqraven** [reply]
i'm running helix 25.01.1 and `.luarc.json` doesn't work for me either. It's working for neovim though so i'm not sure what is the cause. I'm trying to point `workspace.library` to a folder with type annotations and nvim can read `.luarc` and discover that folder without issue

![image](https://github.com/user-attachments/assets/77f12b13-e220-45ce-9545-b0ea85c18cc4)
```json
{
	"$schema": "https://raw.githubusercontent.com/LuaLS/vscode-lua/master/setting/schema.json",
	"workspace.library": ["./wezterm", "../../lib/wezterm-types/types"],
	"runtime.version": "Lua 5.3",
	"hint.enable": true,
	"format.defaultConfig": {
		"indent_style": "space",
		"indent_size": "3",
		"insert_final_newline": "true",
		"space_around_table_field_list": "true",
		"align_continuous_assign_statement": "false"
	}
}
```
both should also be using the same `lua-language-server`
![image](https://github.com/user-attachments/assets/00298815-c766-4a32-a894-2986c7b1dc54)

**Reply by mnpqraven** [reply]
update: helix is now showing types ONLY when i have the files themselves open in the buffer, so i guess helix can't source the paths in `workspace.library` ? this is still pretty unusable because
1. I still have to open the type files which defeat the purpose
2. If the types are split into multiple files then i have to open all those files in the buffer 
![image](https://github.com/user-attachments/assets/be2008ba-05e2-42d9-b987-e8f282f11b9b)

**Reply by CalebLarsen** [reply]
I'm having a hard time reproducing this error @mnpqraven. 

I tried using a similar config and was able to immediately have type information available.

https://github.com/user-attachments/assets/b2f8b9c8-9b03-4374-b29b-e2e70d010acc

Is there something I'm missing on the difference between our setups?

@TurtleP I will add a patch to enable formatting by default to lua. In the meantime, you can add this to your `languages.toml` to get autoformatting
```toml
[[language]]
name = "lua"
injection-regex = "lua"
scope = "source.lua"
file-types = ["lua", "rockspec"]
shebangs = ["lua", "luajit"]
roots = [".luarc.json", ".luacheckrc", ".stylua.toml", "selene.toml", ".git"]
comment-token = "--"
block-comment-tokens = { start = "--[[", end = "--]]" }
indent = { tab-width = 2, unit = "  " }
auto-format = true
language-servers = [ "lua-language-server" ]
```
Be aware that per the [Lua Language Server docs](https://luals.github.io/wiki/formatter/), configuration for indentation should be done at the editor level. For helix, this means either changing the setting in your `languages.toml` for global configuration or in a `.editorconfig` file for local configuration.

**Reply by mnpqraven** [reply]
hi @CalebLarsen , i've managed to finally get it working. I had to put the `.luarc.json` at the root of the workspace otherwise it wouldn't work. My previous workspace was kinda like this. 

```
.git # root
.config/
  wezterm/
    .luarc.json
    other_lua_files.lua
lib/
  lua_library/
```

**Reply by mnpqraven** [reply]
now stylua formatting is not working with the [default configuration](--stdin-filepath), whew...

edit: i fixed it right after posting this (i guess you just need to complain more), needed to have this extra flag
```toml
# languages.toml
formatter = { command = "stylua", args = ["--search-parent-directories", "-"] }
```
